### PR TITLE
guix: Change manifest to use channels and inferiors

### PR DESCRIPTION
**Pull request by dongcarl**
```
This change pins the packages that make up our Guix environment, making
them reproducible regardless of the current version of Guix that the
user is running. It should eliminate the `guix pull' step of the Guix
build process.
```

-----

Not ready for merge yet, as we're waiting on https://issues.guix.gnu.org/issue/36777 to be resolved. The issue is that inferiors don't work for a certain range of Guix versions, which is a minor issues, but should be addressed before we do this.

Also, this currently relies on my repo located here: https://github.com/dongcarl/bitcoin-guix, which is fine because we do similar things in other places, but I'd like to move this repo under the `bitcoin-core` org if at all possible.

### CODE REVIEW

**Comment by fanquake** [comment]
Concept ACK

**Comment by laanwj** [comment]
Concept ACK

**Comment by laanwj** [comment]
> Also, this currently relies on my repo located here: https://github.com/dongcarl/bitcoin-guix, which is fine because we do similar things in other places, but I'd like to move this repo under the bitcoin-core org if at all possible.

Could it be part of [bitcoin-core/packaging](https://github.com/bitcoin-core/packaging), which is the umbrella repository for all things distribution packaging, or does this really need its own separate repo?

**Comment by MarcoFalke** [comment]
According to https://guix.gnu.org/manual/en/html_node/Channels.html

> A channel specifies a URL and branch of a Git repository to be deployed

So yes, it could be a branch in the https://github.com/bitcoin-core/packaging git repo

**Comment by dongcarl** [comment]
It doesn't need to be its own separate repo. Marco is right, we can do branches or even a subdir. I'm not particular about what exactly to do so either of you can make a call and I'll make it happen.

I'll update this accordingly.

**Comment by MarcoFalke** [comment]
> subdir (as long as it's in a git repository)

So it could be put in either https://github.com/bitcoin/bitcoin or https://github.com/bitcoin-core/packaging/. Shouldn't matter much, I think?

**Comment by dongcarl** [comment]
This PR really shines once it can eliminate the `guix pull` step of the Guix build process, this is currently not possible because the last release of Guix did not have a command-line flag that I intend to use (unfortunately you can't use inferiors to travel in time to pick up new command-line flags... yet...).

TLDR: Waiting for new Guix release, after which this will be very useful.

**Comment by dongcarl** [comment]
Closed in favor of #17933
